test/Microsoft.DotNet.Docker.Tests/DockerHelper.cs

```diff
@@ -0,0 +1,105 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using Xunit.Abstractions;
+
+namespace Microsoft.DotNet.Docker.Tests
+{
+    public class DockerHelper
+    {
+        private string DockerOS => GetDockerOS();
+        public string ContainerWorkDir => IsLinuxContainerModeEnabled ? "/sandbox" : "c:\\sandbox";
+        public bool IsLinuxContainerModeEnabled => string.Equals(DockerOS, "linux", StringComparison.OrdinalIgnoreCase);
+        private ITestOutputHelper Output { get; set; }
+
+        public DockerHelper(ITestOutputHelper output)
+        {
+            Output = output;
+        }
+
+        public void Build(string dockerfilePath, string fromImage, string tag, string buildArgs)
+        {
+            string dockerfileContents = File.ReadAllText(dockerfilePath);
+            dockerfileContents = dockerfileContents.Replace("$base_image", fromImage);
+            dockerfilePath = dockerfilePath + ".temp";
+            File.WriteAllText(dockerfilePath, dockerfileContents);
+
+            try
+            {
+                Execute($"build -t {tag} {buildArgs} -f {dockerfilePath} .");
+            }
+            finally
+            {
+                File.Delete(dockerfilePath);
+            }
+        }
+
+        public void DeleteImage(string name)
+        {
+            if (ResourceExists("image", name))
+            {
+                Execute($"image rm -f {name}");
+            }
+        }
+
+        public void DeleteVolume(string name)
+        {
+            if (ResourceExists("volume", name))
+            {
+                Execute($"volume rm -f {name}");
+            }
+        }
+
+        private void Execute(string args)
```

Comment:
How about an Execute method that can return exit code and output text.

```C#
private (int ExitCode, string StandardOutput) Execute(string args, bool isReturnOutputText = false)
        {
            Output.WriteLine($"Executing : docker {args}");

            ProcessStartInfo info = new ProcessStartInfo("docker", args)
            {
                RedirectStandardOutput = true,
                RedirectStandardError = true
            };

            Process process = Process.Start(info);
            process.WaitForExit();
            int exitCode = process.ExitCode;
            Output.WriteLine(process.StandardOutput.ReadToEnd());

            if (exitCode != 0)
            {
                string stdErr = process.StandardError.ReadToEnd();
                string msg = $"Failed to execute {info.FileName} {info.Arguments}{Environment.NewLine}{stdErr}";
                throw new InvalidOperationException(msg);
            }

            string outputText = string.Empty;

            if (isReturnOutputText)
            {
                outputText = process.StandardOutput.ReadToEnd();
            }

            return (exitCode, outputText);
        }
```

This would give us some flexibility. For example, to get Docker version, we would do -

```C#
Output.WriteLine(Execute("version", true).StandardOutput);
```

For this to work we will take a dependency on System.ValueTuple. Not sure what complexity it would add. But wanted to share the thought...

Reply:
I'm not crazy about the usage of Tuples here.